### PR TITLE
Add GitInformation helper.

### PR DESCRIPTION
**Pull request by lgolding**
`GitInformation` and helper classes are from @michaelcfanning, tweaked and tests added by @lgolding. 

Integrated with `InsertOptionalDataVisitor`; introduce `--insert VersionControlInformation` option.

### CODE REVIEW

File: src/Sarif/Core/ArtifactLocation.cs [hunk 1]
```diff
@@ -106,5 +106,30 @@ public ArtifactLocation Resolve(Run run)
                 ? run.Artifacts[Index].Location
                 : this;
         }
+
+        /// <summary>
```
Comment (lgolding) [inline src/Sarif/Core/ArtifactLocation.cs:110]: >/// [](start = 8, length = 3)

This was an extension method, but it doesn't need to be because we have partial clases.

File: src/Test.FunctionalTests.Sarif/MultitoolCommandLineTests.cs [hunk 2]
```diff
@@ -4,10 +4,9 @@
 using System.Diagnostics;
 using System.IO;
 using FluentAssertions;
-using Microsoft.CodeAnalysis.Sarif;
 using Xunit;
 
-namespace Microsoft.CodeAnalysis.Test.FunctionalTests.Sarif
+namespace Microsoft.CodeAnalysis.Sarif.Multitool
```
Comment (lgolding) [inline src/Test.FunctionalTests.Sarif/MultitoolCommandLineTests.cs:9]: >Multitool [](start = 39, length = 9)

While I'm in here, clean up a few namespaces to conform to code base standard that test classes live in the same namespace as corresponding product code.

I did not make the same change in Test.UnitTests.Sarif, even though I added a test file there as well, because there would have been too many changes for this PR.

File: src/Sarif/GitInformation.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.Sarif
```
Comment (lgolding) [inline src/Sarif/GitInformation.cs:9]: >Sarif [](start = 33, length = 5)

I did not put this in a `Git` subdirectory/namespace, since the other three files (IConsoleCapture.cs, ConsoleStreamCapture.cs, and ExternalProcess) are non Git-specific, so this would have been the only type in the namespace.

File: src/Sarif/ExternalProcess.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            psi.CreateNoWindow = true;
+            psi.UseShellExecute = false;
+            psi.RedirectStandardInput = true;
+            psi.RedirectStandardError = true;
+            psi.RedirectStandardOutput = true;
+            psi.StandardOutputEncoding = System.Text.Encoding.UTF8;
+            psi.WindowStyle = ProcessWindowStyle.Hidden;
+
+            using (Process process = Process.Start(psi))
+            using (ManualResetEvent mreOut = new ManualResetEvent(false), mreErr = new ManualResetEvent(false))
```
Comment (lgolding) [inline src/Sarif/ExternalProcess.cs:43]: >mreErr [](start = 74, length = 6)

LGTM complains that these two assignments are useless because neither `mreOut` nor `mreErr` is used. Is there a reason to created these events?

File: src/Sarif/Core/ArtifactLocation.cs (hunk 1 above)
Comment (michaelcfanning) [reply src/Sarif/Core/ArtifactLocation.cs:110]: Right but does this functionality conceptually belong as part of ArtifactLocation? I'm not sure there's a good case for it.

---
In reply to: [472406181](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r472406181) [](ancestors = 472406181)

File: src/Sarif/Visitors/InsertOptionalDataVisitor.cs [hunk 5]
```diff
@@ -62,6 +68,28 @@ public override Run VisitRun(Run node)
             return visited;
         }
 
+        private void InsertVersionControlInformation()
+        {
+            var gitInformation = new GitInformation(s_fileSystem);
+            string currentDirectory = s_executionEnvironment.CurrentDirectory;
```
Comment (michaelcfanning) [inline src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: >CurrentDirectory [](start = 61, length = 16)

This will work some of the time but seems haphazard. What we need to happen is for every physical location that is included in the log file be examined for relevance to source control.

File: src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/ExpectedOutputs/CoreTests-Absolute_TextFiles.sarif [hunk 6]
```diff
@@ -65,7 +65,7 @@
       "artifacts": [
         {
           "location": {
-            "uri": "file:///REPLACED_AT_TEST_RUNTIME/src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/InsertOptionalDataVisitor.txt"
+            "uri": "file:///ENLISTMENT_ROOT/src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/InsertOptionalDataVisitor.txt"
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/ExpectedOutputs/CoreTests-Absolute_TextFiles.sarif:68]: >ENLISTMENT_ROOT [](start = 28, length = 15)

Why did you change this name? It was named previously as a clue that this value is rewritten. We've lost that information now.

Or did you discover that the renaming no longer occurs?

File: src/Test.UnitTests.Sarif/Visitors/InsertOptionalDataVisitorTests.cs [hunk 7]
```diff
@@ -26,22 +27,32 @@ public InsertOptionalDataVisitorTests(ITestOutputHelper outputHelper, DeletesOut
...
                 GetResourceText(inputResourceName),
                 formatting: Formatting.Indented,
-                out string transformedLog);
+                updatedLog: out _);
+
+            // Some of the tests operate on SARIF files that mention the absolute path of the file
+            // that was "analyzed" (InsertOptionalDataVisitor.txt). That path depends on the repo
+            // root, and so can vary depending on the machine where the tests are run. To avoid
+            // this problem, both the input files and the expected output files contain a fixed
+            // string "ENLISTMENT_ROOT" in place of the directory portion of the path. But some
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif/Visitors/InsertOptionalDataVisitorTests.cs:39]: >ENLISTMENT_ROOT [](start = 23, length = 15)

Can you please think of a name that is an improvement on the previous (that meets the standard that prompted you to change it) but which still indicates that the value is rewritten by the test infrastructure? This is not obvious after the change you've made.

File: src/Sarif/ExternalProcess.cs (hunk 4 above)
Comment (michaelcfanning) [reply src/Sarif/ExternalProcess.cs:43]: This is just dead code from an earlier draft, you can remove it.

---
In reply to: [472434440](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r472434440) [](ancestors = 472434440)

File: src/Sarif/Visitors/InsertOptionalDataVisitor.cs (hunk 5 above)
Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: `ENVIRONMENT_DEPENDENT_ENLISTMENT_ROOT`?

---
In reply to: [475661249](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475661249) [](ancestors = 475661249)

File: src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/ExpectedOutputs/CoreTests-Absolute_TextFiles.sarif (hunk 6 above)
Comment (lgolding) [reply src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/ExpectedOutputs/CoreTests-Absolute_TextFiles.sarif:68]: `ENVIRONMENT_DEPENDENT_ENLISTMENT_ROOT`?

---
In reply to: [475662994](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475662994) [](ancestors = 475662994)

Comment (michaelcfanning) [reply src/Test.UnitTests.Sarif/TestData/InsertOptionalDataVisitor/ExpectedOutputs/CoreTests-Absolute_TextFiles.sarif:68]: That doesn't indicate to readers that the test runtime replaces this value. Can you please just restore it to its previous value?

---
In reply to: [475669236](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475669236) [](ancestors = 475669236,475662994)

File: src/Sarif/Visitors/InsertOptionalDataVisitor.cs (hunk 5 above)
Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: This is the directory from which `GitInformation.GetRepositoryRoot` starts its upward search for the directory containing `.git`. The assumption was that this operation works only if you're running from somewhere under the repo root.

**BUT**: Now I see in `VisitPhysicalLocation` below, where it is enriching the regions and filling in the snippets, that it relies on `originalUriBaseIds` to resolve URIs -- and it if can't resolve a URI it silently fails to fill in the information.

So how about this: We try the operation starting with each entry in `originalUriBaseIds`. If all of them fail (that is, if there is no entry that points at or into a directory containing a `.git`), only then do we throw. (And we _do_ throw; I don't think it's ok to silently do nothing here.) We have to be careful in the presence of submodules to pick the topmost directory with a `.git`.

---
In reply to: [475661249](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475661249) [](ancestors = 475661249)

Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: (Having said that, I don't understand what you mean by examining every physical location. Do you mean "See if any physical location mentioned anywhere in the log file is an absolute URI from which the repo root can be determined?")

---
In reply to: [475691019](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475691019) [](ancestors = 475691019,475661249)

Comment (michaelcfanning) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: Yes. The goal of this feature is that any physical location referenced in the log file is expressed as a relative path with corresponding enlistment root, if that PLC is in fact a source-controlled item. The code as authored is not accomplishing this. 

---
In reply to: [475692378](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475692378) [](ancestors = 475692378,475691019,475661249)

Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: There are two distinct issues here:

1. How do we determine the repo root (and thus populate `run.versionControlProvenance`)? My implementation relies on the current directory; my alternative proposal above would rely on `originalUriBaseIds`, similar to what `VisitPhysicalLocation` does below.

2. Having determined the repo root, how do we express all URIs with respect to it? We're 100% agreed that we need to do that. **BUT**: as the Multitool is currently structured, that's not the job of the `rewrite` command/`InsertOptionalDataVisitor`. Rather, it's the job of the `rebaseuri` command/`RebaseUriVisitor`.

So IMO the code is structured correctly, and what the normalizer should do is invoke both the `rewrite` command and the `rebaseuri` command -- **EXCEPT** it turns out that the `rebaseuri` command supports the options for, and invokes, the `InsertOptionalDataVisitor`, and so it's actually a superset of the `rewrite` command, so a single invocation of `rebaseuri` will suffice.

This gets into the whole multitool redesign discussion. We can choose to do it now, or after the normalizer is done. But either way, the URI rebasing doesn't belong in the `InsertOptionalDataVisitor`.

---
In reply to: [475696046](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475696046) [](ancestors = 475696046,475692378,475691019,475661249)

Comment (michaelcfanning) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: I suggest we get this code in now. As we discussed offline, commands like rebaseuri are redudant and more granular than we need. 

I will put some thought into the larger design.

Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: Ok, I will merge what I have.

---
In reply to: [475714891](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475714891) [](ancestors = 475714891)

Comment (lgolding) [reply src/Sarif/Visitors/InsertOptionalDataVisitor.cs:74]: I mean, "Ok, please merge what I have."

---
In reply to: [475769205](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475769205) [](ancestors = 475769205,475714891)
